lib/webpush/encryption.rb

```diff
@@ -34,10 +36,15 @@ def encrypt(message, p256dh, auth)
       nonce = HKDF.new(prk, salt: salt, info: nonce_info).next_bytes(12)
 
       ciphertext = encrypt_payload(message, content_encryption_key, nonce)
+      # This should never happen if message length <= 3993
+      raise ArgumentError, "encrypted payload is too big" if ciphertext.bytesize > 4096
 
       serverkey16bn = convert16bit(server_public_key_bn)
-      rs = ciphertext.bytesize
-      raise ArgumentError, "encrypted payload is too big" if rs > 4096
+      # According to RFC8188, the final record can be smaller than the record size
+      # RFC8291 requires encrypted messages to be at most 4096. And the example set the
+      # RS to 4096. RS can be smaller to 4096 but hardcoding is also following the specs.
+      # We set RS=4096 to allow testing with the RFC example.
+      rs = 4096
```

Comment:
This is one thing I'm a bit confused by, though I'm not really knowledgeable about aes128gcm. Why is `rs` now unrelated to the actual record size?

Reply:
RS gives the maximum size of the record, RFC8291 tells to use 4096 (it should be possible to use a lower rs, but it isn't necessary, and breaks the example. It may be possible that this rs was the cause of the decryption failure with tink lib)

Reply:
> An application server MUST set the "rs" parameter in the "aes128gcm" content coding header to __a size that is greater than__ the sum of the lengths of the plaintext, the padding delimiter (1 octet), any padding, and the authentication tag (16 octets).

It means it can always be 4096, but may be lower. That is what they do with the example:

>    The salt, record size of 4096, and application server public key
   produce an 86-octet header of:
>
>   DGv6ra1nlYgDCS1FRnbzlwAAEABBBP4z 9KsN6nGRTbVYI_c7VJSPQTBtkgcy27ml
   mlMoZIIgDll6e3vCYLocInmYWAmS6Tlz AC8wEqKK6PBru3jl7A8

And being able to reproduce the example is IMO very important for a library implementing an RFC